src/manifolds/Decorator.jl

```diff
@@ -0,0 +1,578 @@
+    error(manifold_function_not_implemented_message(M, get_basis, p, B))
+end
+
+
+function hat!(M::DT, X, p, Xⁱ) where {DT<:AbstractDecoratorManifold}
+    return hat!(M, X, p, Xⁱ, _acts_transparently(M, hat!))
+end
+function hat!(M::DT, X, p, Xⁱ, ::Val{true}) where {DT<:AbstractDecoratorManifold}
+    return hat!(M.manifold, X, p, Xⁱ)
+end
+function hat!(M::DT, X, p, Xⁱ, ::Val{false}) where {DT<:AbstractDecoratorManifold}
+    error(manifold_function_not_implemented_message(M, hat!, p, X))
+end
+
+function injectivity_radius(M::DT) where {DT<:AbstractDecoratorManifold}
+    return injectivity_radius(M, _acts_transparently(M, injectivity_radius))
+end
+function injectivity_radius(M::DT, ::Val{true}) where {DT<:AbstractDecoratorManifold}
+    return injectivity_radius(M.manifold)
+end
+function injectivity_radius(M::DT, ::Val{false}) where {DT<:AbstractDecoratorManifold}
+    error(manifold_function_not_implemented_message(M, injectivity_radius))
+end
+function injectivity_radius(M::DT, p) where {DT<:AbstractDecoratorManifold}
+    return injectivity_radius(M, p, _acts_transparently(M, injectivity_radius))
+end
+function injectivity_radius(M::DT, p, ::Val{true}) where {DT<:AbstractDecoratorManifold}
+    return injectivity_radius(M.manifold, p)
+end
+function injectivity_radius(M::DT, p, ::Val{false}) where {DT<:AbstractDecoratorManifold}
+    error(manifold_function_not_implemented_message(M, injectivity_radius, p))
+end
+
+function inner(M::DT, p, X, Y) where {DT<:AbstractDecoratorManifold}
+    return inner(M, p, X, Y, _acts_transparently(M, inner))
+end
+function inner(M::DT, p, X, Y, ::Val{true}) where {DT<:AbstractDecoratorManifold}
+    return inner(M.manifold, p, X, Y)
+end
+function inner(M::DT, p, X, Y, ::Val{false}) where {DT<:AbstractDecoratorManifold}
+    error(manifold_function_not_implemented_message(M, inner, p, X, Y))
+end
+
+function inverse_retract!(
+    M::DT,
+    X,
+    p,
+    q,
+    m::AbstractInverseRetractionMethod
+) where {DT<:AbstractDecoratorManifold}
+    return inverse_retract!(M, X, p, q, m, _acts_transparently(M, inverse_retract!))
+end
+function inverse_retract!(
+    M::DT,
+    X,
+    p,
+    q,
+    m::LogarithmicInverseRetraction
+) where {DT<:AbstractDecoratorManifold}
+    return inverse_retract!(M, X, p, q, m, _acts_transparently(M, inverse_retract!))
+end
+function inverse_retract!(
+    M::DT,
+    X,
+    p,
+    q,
+    m::AbstractInverseRetractionMethod,
+    ::Val{true}
+) where {DT<:AbstractDecoratorManifold}
+    return inverse_retract!(M.manifold, X, p, q, m)
+end
+function inverse_retract!(
+    M::DT,
+    X,
+    p,
+    q,
+    m::AbstractInverseRetractionMethod,
+    ::Val{false}
+) where {DT<:AbstractDecoratorManifold}
+    error(manifold_function_not_implemented_message(M, inverse_retract!, X, p, q, m))
+end
+
+function isapprox(M::DT, p, q) where {DT<:AbstractDecoratorManifold}
```

Comment:
This needs keyword arguments for tolerance.

Reply:
Yes, that's maybe copy paste, but this file will simplify with your macro tremendously anyways, then I'll add them. You can push to this branch directly if you want, I am working for the next 30 minutes on a final comparison for the logo.

Reply:
I think I'll leave further work on this for tomorrow.

Reply:
I can help redoing `Decorator.jl` when you are finished doing the macro. Once this is done we should wait for Seth and if he agrees move the basic stuff even to base (decorator, macro and the decoration of the basic interface functions).